exec/java-exec/src/main/java/org/apache/drill/exec/planner/common/DrillStatsTable.java

```diff
@@ -0,0 +1,161 @@
+import org.apache.drill.exec.proto.UserBitShared;
+import org.apache.drill.exec.server.DrillbitContext;
+import org.apache.drill.exec.server.rest.QueryWrapper.Listener;
+
+/**
+ * Wraps the stats table info including schema and tableName. Also materializes stats from storage and keeps them in
+ * memory.
+ */
+public class DrillStatsTable {
+  private static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(DrillStatsTable.class);
+
+  /**
+   * List of columns in stats table.
+   */
+  public static final String COL_COLUMN = "column";
+  public static final String COL_COMPUTED = "computed";
+  public static final String COL_STATCOUNT = "statcount";
+  public static final String COL_NDV = "ndv";
+
+  private final String schemaName;
+  private final String tableName;
+
+  private final Map<String, Long> ndv = Maps.newHashMap();
+  private double rowCount = -1;
+
+  private boolean materialized = false;
+
+  public DrillStatsTable(String schemaName, String tableName) {
+    this.schemaName = schemaName;
+    this.tableName = tableName;
+  }
+
+  /**
+   * Get number of distinct values of given column. If stats are not present for the given column, a null is returned.
+   *
+   * Note: returned data may not be accurate. Accuracy depends on whether the table data has changed after the
+   * stats are computed.
+   *
+   * @param col
+   * @return
+   */
+  public Double getNdv(String col) {
+    Preconditions.checkState(materialized, "Stats are not yet materialized.");
+
+    final String upperCol = col.toUpperCase();
+    final Long ndvCol = ndv.get(upperCol);
+    if (ndvCol != null) {
+      return Math.min(ndvCol, rowCount);
+    }
+
+    return null;
+  }
+
+  /**
+   * Get row count of the table. Returns null if stats are not present.
+   *
+   * Note: returned data may not be accurate. Accuracy depends on whether the table data has changed after the
+   * stats are computed.
+   *
+   * @return
+   */
+  public Double getRowCount() {
+    Preconditions.checkState(materialized, "Stats are not yet materialized.");
+    return rowCount > 0 ? rowCount : null;
+  }
+
+  /**
+   * Read the stats from storage and keep them in memory.
+   * @param context
+   * @throws Exception
+   */
+  public void materialize(final QueryContext context) throws Exception {
+    if (materialized) {
+      return;
+    }
+
+    final String fullTableName = "`" + schemaName + "`.`" + tableName + "`";
+    final String sql = "SELECT a.* FROM " + fullTableName + " AS a INNER JOIN " +
+        "(SELECT `" + COL_COLUMN + "`, max(`" + COL_COMPUTED +"`) AS `" + COL_COMPUTED + "` " +
+        "FROM " + fullTableName + " GROUP BY `" + COL_COLUMN + "`) AS b " +
+        "ON a.`" + COL_COLUMN + "` = b.`" + COL_COLUMN +"` and a.`" + COL_COMPUTED + "` = b.`" + COL_COMPUTED + "`";
+
+    final DrillbitContext dc = context.getDrillbitContext();
+    try(final DrillClient client = new DrillClient(dc.getConfig(), dc.getClusterCoordinator(), dc.getAllocator())) {
```

Comment:
Few questions here:
(1) How does this work when authentication is enabled?
(2) Why does this use a listener from the rest package?
(3) Are there alternative way to run this query?